src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandleMetadata.cs

```diff
@@ -3,49 +3,216 @@
+                    if (converter.ClassType == ClassType.Enumerable)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataPreservedArrayValuesNotFound(converter.TypeToConvert);
+                    }
+
+                    // The reader should have detected other invalid cases.
+                    Debug.Assert(reader.TokenType == JsonTokenType.EndObject);
+
+                    value = default!;
+                    return true;
+                }
+
+                ReadOnlySpan<byte> propertyName = GetSpan(ref reader);
+                MetadataPropertyName metadata = GetMetadataPropertyName(propertyName);
+                if (metadata == MetadataPropertyName.Id)
+                {
+                    state.Current.JsonPropertyName = propertyName.ToArray();
+                    if (!converter.CanHaveIdMetadata)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataCannotParsePreservedObjectIntoImmutable(converter.TypeToConvert);
+                    }
+
+                    state.Current.ObjectState = StackFrameObjectState.MetadataIdProperty;
+                }
+                else if (metadata == MetadataPropertyName.Ref)
+                {
+                    state.Current.JsonPropertyName = propertyName.ToArray();
+                    if (converter.TypeToConvert.IsValueType)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataInvalidReferenceToValueType(converter.TypeToConvert);
+                    }
+
+                    state.Current.ObjectState = StackFrameObjectState.MetadataRefProperty;
+                }
+                else if (metadata == MetadataPropertyName.Values)
+                {
+                    state.Current.JsonPropertyName = propertyName.ToArray();
+                    if (converter.ClassType == ClassType.Enumerable)
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataMissingIdBeforeValues();
+                    }
+                    else
+                    {
+                        ThrowHelper.ThrowJsonException_MetadataInvalidPropertyWithLeadingDollarSign(propertyName, ref state, reader);
+                    }
+                }
+                else
+                {
+                    Debug.Assert(metadata == MetadataPropertyName.NoMetadata);
+
+                    // Having a StartObject without metadata properties is not allowed.
+                    if (converter.ClassType == ClassType.Enumerable)
+                    {
+                        state.Current.JsonPropertyName = propertyName.ToArray();
+                        ThrowHelper.ThrowJsonException_MetadataPreservedArrayInvalidProperty(converter.TypeToConvert, reader, ref state);
+                    }
+
+                    // Skip the read of the first property name, since we already read it above.
+                    state.Current.PropertyState = StackFramePropertyState.ReadName;
```

Comment:
Could you provide `MyStream` class?

Reply:
```C#
private class MyStream : Stream
{
    private int _index;
    private readonly byte[] _data;
    public MyStream(string json)
    {
        _index = 0;
        _data = Encoding.UTF8.GetBytes(json);
    }
    public override bool CanRead => true;
    public override bool CanSeek => throw new NotImplementedException();
    public override bool CanWrite => throw new NotImplementedException();
    public override long Length => _data.Length;
    public override long Position { get => throw new NotImplementedException(); set => throw new NotImplementedException(); }
    public override void Flush()
    {
        throw new NotImplementedException();
    }
    public override int Read(byte[] buffer, int offset, int count)
    {
        int min = Math.Min(_data.Length - _index, count);
        _data.AsSpan(_index, min).CopyTo(buffer.AsSpan(offset, count));
        _index += min;
        return min;
    }
    public override long Seek(long offset, SeekOrigin origin)
    {
        throw new NotImplementedException();
    }
    public override void SetLength(long value)
    {
        throw new NotImplementedException();
    }
    public override void Write(byte[] buffer, int offset, int count)
    {
        throw new NotImplementedException();
    }
}
```

Reply:
@Jozkee please extend the existing stream read\write tests to include types that use reference handling.

I believe we just need this line to set the state, so we don't re-enter ResolveMetadata on success of reading the property name:
```cs
                if (reader.TokenType != JsonTokenType.PropertyName)
                {
...
                    // The reader should have detected other invalid cases.
                    Debug.Assert(reader.TokenType == JsonTokenType.EndObject);

                    // Skip the read of the first property name, since we already read it above.
>> here                    state.Current.PropertyState = StackFramePropertyState.ReadName;
before returning `true`
```